src/store/products/products.types.ts

```diff
@@ -0,0 +1,45 @@
+// Entities
+
+export type ProductId = number;
+
+export interface Product {
+  product_id: ProductId;
+  name: string;
+  purchase_cost: number;
+}
+
+// Action types
+
+export const PRODUCTS_FETCH_SUCCESS = "@@products/FETCH_SUCCESS";
+export const PRODUCTS_FETCH_BY_ID_SUCCESS = "@@products/FETCH_BY_ID_SUCCESS";
+export const PRODUCTS_ADD_SUCCESS = "@@products/ADD_SUCCESS";
+
+// Products Api
+
+export interface ProductsFetchResponseData {
+  products: Product[];
+}
+export interface ProductsAddParams {
+  name: string;
+}
+export interface ProductsAddResponseData {
+  product: Product;
+}
+
+// Actions
+
+interface ProductsFetchSuccess {
+  type: typeof PRODUCTS_FETCH_SUCCESS;
```

Comment:
why need it say 'typeof'? We already said PRODUCTS_FETCH_SUCCESS is a type..

Reply:
Yeah. There's some confusing terminology going on. Sorry. 
This is because we have typescript types, but also the unique identifier for an action is called the action type.

The redux typescript stuff is a bit weird. I'd try not to worry too much about it for now.

https://redux.js.org/recipes/usage-with-typescript

It's basically so that the action types and payloads can be typesafe. i.e. if you tried to dispatch an action and the `action.type` and the `action.payload` weren't consistent with eachother, then the code wouldn't compile.

Reply:
OK I'll just roll with that for now.